lib/util.py

```diff
@@ -678,79 +679,135 @@ class TxHashMismatch(ServerError):
-        ''' Raised by get() if max_message_bytes is set and the message size
-        limit was exceeded. '''
+    class Closed(RuntimeError):
+        ''' Raised if socket is closed '''
 
     def __init__(self, socket, *, max_message_bytes=0):
         ''' A max_message_bytes of <= 0 means unlimited, otherwise a positive
         value indicates this many bytes to limit the message size by. This is
         used by get(), which will raise MessageSizeExceeded if the message size
         received is larger than max_message_bytes. '''
         self.socket = socket
-        self.message = b''
-        self.set_timeout(0.1)
+        socket.settimeout(0)
         self.recv_time = time.time()
         self.max_message_bytes = max_message_bytes
-
-    def set_timeout(self, t):
-        self.socket.settimeout(t)
+        self.recv_buf = bytearray()
+        self.send_buf = bytearray()
 
     def idle_time(self):
         return time.time() - self.recv_time
 
-    def clean_up(self):
-        ''' Clears the receive buffer to make sure no garbage data remains '''
-        self.message = b''
+    def get_selectloop_info(self):
+        ''' Returns tuple:
+
+        read_pending - new data is available that may be unknown to select(),
+            so perform a get() regardless of select().
+        write_pending - some send data is still buffered, so make sure to call
+            send_flush if writing becomes available.
+        '''
+        try:
+            # pending() only defined on SSL sockets.
+            has_pending = self.socket.pending() > 0
+        except AttributeError:
+            has_pending = False
+        return has_pending, bool(self.send_buf)
 
     def get(self):
+        ''' Attempt to read out a message, possibly saving additional messages in
+        a receive buffer.
+
+        If no message is currently available, this raises util.timeout and you
+        should retry once data becomes available to read. If connection is bad for
+        some known reason, raises .Closed; other errors will raise other exceptions.
+        '''
         while True:
-            response, self.message = parse_json(self.message)
+            response, self.recv_buf = parse_json(self.recv_buf)
             if response is not None:
                 return response
+
             try:
                 data = self.socket.recv(1024)
-            except socket.timeout:
+            except (socket.timeout, BlockingIOError, ssl.SSLWantReadError):
                 raise timeout
-            except ssl.SSLError:
-                raise timeout
-            except socket.error as err:
-                if err.errno == 60:
-                    raise timeout
-                elif err.errno in [11, 35, 10035]:
-                    self.print_error("socket errno %d (resource temporarily unavailable)"% err.errno)
-                    time.sleep(0.2)
+            except OSError as exc:
+                if exc.errno in (11, 35, 60, 10035):
```

Comment:
Before this code "accepted" `errno in (11, 35, 10035)`, now it "rejects" those errno codes.  Any reason for this change ?

I see actually that raising `timeout` is not so bad -- is that correct?  It just leads to "passing" on this pipe this time around.. is that right? So this is functionally equivalent to before, more or less, without the hard-inserted `time.sleep` here?  Is that right?

Sorry -- this code is just so spaghetti... I had to ask.

Reply:
Hmm I think before/after it raises timeout for those codes, the only difference being that I got rid of the sleep (which is not needed -- that's why we have the select).

Reply:
Ah ok. My bad.  

Why did the original code horribly sleep there, I wonder?

Reply:
Well it was 28df27fba2 and ab1714e9f2 that put in the sleep (and incremented it)... no real reason. I think in olden days there wasn't even a select loop to begin with, so that probably was a contributing factor.

Reply:
Yowza.  Ok, I'm satisfied and trust your judgement here.  Not sure what's going on TBH but sleeping randomly there just looks wrong anyway and is unneeded it seems so.. I'm satisfied.